### PR TITLE
[FIRRTLToHW] Simplify lowerRegConnect, support HW wires

### PR DESCRIPTION
**Pull request by fabianschuiki**
Simplify the `lowerRegConnect` function such that it rejects subfield, subindex, and subaccesses on its left hand side. Consensus after some discussion in #4826 is that any subaccesses on the left hand side of a connect should be handled in a pre-pass that refactors the assignment in the exact way that `lowerRegConnect` used to do until now. This significantly simplifies the implementation of `LowerToHW` and moves the complexities of sink subaccesses into a dedicated pass.

Rename the `lowerRegConnect` function to `lowerConnect` and make it work with `hw::WireOp` in addition to `seq::FirRegOp`.

This removes the need to carry a separate `regMapping` for the backedges used as the `next` value of registers. Instead, `lowerConnect` directly inspects the operation on the LHS of the connect and does the right thing if it's a `seq::FirRegOp` or `hw::WireOp`.

The lowering of FIRRTL connect and strictconnect now first attempt to materialize connects in wires and regs directly through `lowerConnect`, and only if that's not possible (e.g. because the LHS is not a `seq::FirRegOp` or `hw::WireOp`) does it resort to generating assign ops.

### CODE REVIEW

**Review by uenoku — COMMENTED** [review]
I think a lot of complexity come from sink subaccess lowering. Do we want to handle it at LowerToHW? I feel it's monolithic to handle everything at LowerToHW (probably dynamic index ArrayInjectOp would be rescue though).

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp [hunk 1]
```diff
@@ -3696,94 +3721,194 @@ LogicalResult FIRRTLLowering::visitStmt(SkipOp op) {
...
-  if (fieldID != 0) {
-    nextValue =
-        inject(seqReg.getNext(), regOp->getResult(0).getType(), fieldID);
+    // Handle the case where the index is dynamic. In that case we need to
+    // create a mux for each element that updates the element value if the
+    // accessed index matches.
+    SmallVector<Value> elements;
+    for (unsigned idx = 0, end = ty.getSize(); idx != end; ++idx) {
+      auto idxVal = getOrCreateIntConstant(indexWidth, idx);
+      auto cmpOp = builder.create<comb::ICmpOp>(comb::ICmpPredicate::eq,
```
[the remark this answers is not in the document]
Comment (darthscsi) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp:3888]: LowerToHW almost always should add a bin flag.

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp [hunk 2]
```diff
@@ -3696,94 +3721,194 @@ LogicalResult FIRRTLLowering::visitStmt(SkipOp op) {
...
+                                                op.getIndex(), idxVal);
+      auto oldVal = builder.create<hw::ArrayGetOp>(subaccessInput, idxVal);
+      auto muxOp = builder.create<comb::MuxOp>(cmpOp, srcVal, oldVal);
```
[the remark this answers is not in the document]
Comment (darthscsi) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp:3891]: LowerToHW almost always should add a bin flag.

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp:3888]: Yep my bad, forgot about that. Thanks!

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp:3891]: Yep my bad, forgot about that. Thanks!

**Comment by fabianschuiki** [comment]
> @uenoku I think a lot of complexity come from sink subaccess lowering. Do we want to handle it at LowerToHW? I feel it's monolithic to handle everything at LowerToHW (probably dynamic index ArrayInjectOp would make it less monolithic though).

Yeah I agree. My reasoning here was that `lowerRegConnect` already exists and it does the subaccess lowering. So the delta to making it work for wires (and adding subaccess support) was a lot smaller than stripping it out altogether.

But I agree, ideally we'd guarantee through the FIRRTL pipeline that there are no sink subaccesses anymore when we hit LowerToHW. I don't know how that will work with aggregate preservation. Would we just create this struct/array injects at the FIRRTL dialect level?

My proposal: land this small generalization of `lowerRegConnect` and then look into if we can actually guarantee absence of subaccess sinks when we hit LowerToHW, possibly inserting new passes that do this inject/splicing at the FIRRTL level, and strip out the complexity from LowerToHW. WDYT?

**Comment by uenoku** [comment]
>  I don't know how that will work with aggregate preservation. Would we just create this struct/array injects at the FIRRTL dialect level?

Even with aggregate preservation, there should be no sink subaccess at LowerToHW. There could be source subaccess though. 


> My proposal: land this small generalization of lowerRegConnect and then look into if we can actually guarantee absence of subaccess sinks when we hit LowerToHW, possibly inserting new passes that do this inject/splicing at the FIRRTL level, and strip out the complexity from LowerToHW. WDYT?

Yeah, I think just rejecting sink subaccess looks good to me.

**Comment by fabianschuiki** [comment]
Sounds good @uenoku, let me just rip that part of the code out and replace it with an error if there's a sink subaccess. That'll make the whole thing a lot simpler :100:

**Comment by fabianschuiki** [comment]
@uenoku This PR now just simplifies `lowerRegConnect` into a trivial `lowerConnect` that also handles `hw.wire` and that complains if there are subaccess on the LHS.

**Comment by uenoku** [comment]
> @uenoku This PR now just simplifies lowerRegConnect into a trivial lowerConnect that also handles hw.wire and that complains if there are subaccess on the LHS.

Nice! I tested large cores with aggregate preservation but there was no failure 👍

**Pull request merged**